cpp/src_prims/stats/histogram.h

```diff
@@ -0,0 +1,488 @@
+enum HistType {
+  /** use only global atomics */
+  HistTypeGmem,
+  /**
+   * global mem atomics using warp-level operations sm70.
+   * should work well when the input data is very skewed
+   * will fall back to HistTypeGmem on unsupported arch's
+   */
+  HistTypeGmemWarp,
+  /** uses shared mem atomics to reduce global traffic */
+  HistTypeSmem,
+  /**
+   * shared mem atomics using warp-level operations sm70.
+   * should work well when the input data is very skewed
+   * will fall back to HistTypeSmem on unsupported arch's
+   */
+  HistTypeSmemWarp,
+  /** shared mem atomics but with bins to be 2B int's */
+  HistTypeSmemBits16,
+  /** shared mem atomics but with bins to ba 1B int's */
+  HistTypeSmemBits8,
+  /** shared mem atomics but with bins to be 4b int's */
+  HistTypeSmemBits4,
+  /** shared mem atomics but with bins to be 2b int's */
+  HistTypeSmemBits2,
+  /** shared mem atomics but with bins to be 1b int's */
+  HistTypeSmemBits1,
+  /** builds a hashmap of active bins in shared mem */
+  HistTypeSmemHash,
+  /** decide at runtime the best algo for the given inputs */
+  HistTypeAuto
+};
+
+template <typename DataT, typename BinnerOp, typename IdxT, int VecLen,
+          typename CoreOp>
+DI void histCoreOp(const DataT* data, IdxT n, BinnerOp binOp, CoreOp op) {
+  auto bdim = IdxT(blockDim.x);
+  IdxT tid = threadIdx.x + bdim * blockIdx.x;
+  tid *= VecLen;
+  IdxT stride = bdim * gridDim.x * VecLen;
+  typedef TxN_t<DataT, VecLen> VecType;
+  VecType a;
+  for (auto i = tid; i < n; i += stride) {
+    a.load(data, i);
+#pragma unroll
+    for (int j = 0; j < VecLen; ++j) {
+      int binId = binOp(a.val.data[j], i + j);
+      op(binId, i + j);
+    }
+  }
+}
+
+template <typename DataT, typename BinnerOp, typename IdxT, int VecLen>
+__global__ void gmemHistKernel(int* bins, const DataT* data, IdxT n,
+                               BinnerOp binner) {
+  auto op = [=] __device__(int binId, IdxT idx) { atomicAdd(bins + binId, 1); };
+  histCoreOp<DataT, BinnerOp, IdxT, VecLen>(data, n, binner, op);
+}
+
+template <typename DataT, typename BinnerOp, typename IdxT, int VecLen>
+void gmemHist(int* bins, IdxT nbins, const DataT* data, IdxT n, BinnerOp op,
+              int tpb, cudaStream_t stream) {
+  int nblks = ceildiv<int>(VecLen ? n / VecLen : n, tpb);
+  gmemHistKernel<DataT, BinnerOp, IdxT, VecLen>
+    <<<nblks, tpb, 0, stream>>>(bins, data, n, op);
+  CUDA_CHECK(cudaGetLastError());
+}
+
+template <typename DataT, typename BinnerOp, typename IdxT, int VecLen>
+__global__ void smemHistKernel(int* bins, const DataT* data, IdxT n, IdxT nbins,
+                               BinnerOp binner) {
+  extern __shared__ unsigned sbins[];
+  for (auto i = threadIdx.x; i < nbins; i += blockDim.x) {
+    sbins[i] = 0;
+  }
+  __syncthreads();
+  auto op = [=] __device__(int binId, IdxT idx) {
+    atomicAdd(sbins + binId, 1);
+  };
+  histCoreOp<DataT, BinnerOp, IdxT, VecLen>(data, n, binner, op);
```

Comment:
Processing multiple elements per thread is required to get the benefit from shared memory atomics.

The simplest approach is to always launch a fixed number of thread blocks, e.g. `2 * n_multiprocesors`, and distribute the data among them.

This also applies to other kernels.

Reply:
this should be fixed in the commits I'll be pushing soon. Thanks for pointing it out.